File: app/src/main/java/com/helloelliote/util/retrofit/ProgressBody.java
package com.helloelliote.util.retrofit;

import android.os.Handler;
import android.os.Looper;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;

public final class ProgressBody extends RequestBody {

    private static final int DEFAULT_BUFFER_SIZE = 2048;
    private File file;
    private String contentType;
    private UploadCallback callback;

    public ProgressBody(final File file, String contentType, final UploadCallback callback) {
        this.file = file;
        this.contentType = contentType;
        this.callback = callback;
    }

    @Override
    public long contentLength() throws IOException {
        return file.length();
    }

    @Override
    public MediaType contentType() {
        return MediaType.parse(contentType + "/*");
    }

    @Override
    public void writeTo(@NotNull BufferedSink sink) throws IOException {
        final long totalSize = file.length();
        final byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        try (FileInputStream inputStream = new FileInputStream(file)) {
            long uploadSize = 0L;
            int readSize;
            int number = 0;
            Handler handler = new Handler(Looper.getMainLooper());
            while ((readSize = inputStream.read(buffer)) != -1) {
                int progress = (int) (100 * uploadSize / totalSize);
                if (progress > number + 1) {
                    // update progress on UI thread
                    handler.post(new ProgressUpdater(uploadSize, totalSize));
                    number = progress;
                }
                uploadSize += readSize;
                sink.write(buffer, 0, readSize);
            }
        }
    }

    public interface UploadCallback {
        void onInitiate(int percentage);

        void onProgress(int percentage);

        void onError();

        void onFinish(int percentage);
    }

    private class ProgressUpdater implements Runnable {
        private long uploadSize;
        private long totalSize;

        ProgressUpdater(long uploadSize, long totalSize) {
            this.uploadSize = uploadSize;
            this.totalSize = totalSize;
        }

        @Override
        public void run() {
            callback.onProgress((int) (100 * uploadSize / totalSize));
        }
    }
}

File: app/src/main/java/kr/djspi/pipe01/network/ProgressBody.kt
package kr.djspi.pipe01.network

import android.os.Handler
import android.os.Looper
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okio.BufferedSink
import java.io.File
import java.io.FileInputStream
import java.io.IOException

class ProgressBody(
    private val file: File,
    private val contentType: String,
    private val callback: UploadCallback
) :
    RequestBody() {
    override fun contentLength(): Long {
        return file.length()
    }

    override fun contentType(): MediaType? {
        return "$contentType/*".toMediaTypeOrNull()
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        val totalSize = file.length()
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        FileInputStream(file).use { inputStream ->
            var uploadSize = 0L
            var readSize: Int
            var number = 0
            val handler = Handler(Looper.getMainLooper())
            while (inputStream.read(buffer).also { readSize = it } != -1) {
                val progress = (100 * uploadSize / totalSize).toInt()
                if (progress > number + 1) {
                    // update progress on UI thread

                    handler.post(ProgressUpdater(uploadSize, totalSize))
                    number = progress
                }
                uploadSize += readSize.toLong()
                sink.write(buffer, 0, readSize)
            }
        }
    }

    interface UploadCallback {
        fun onInitiate(percentage: Int)
        fun onProgress(percentage: Int)
        fun onError()
        fun onFinish(percentage: Int)
    }

    private inner class ProgressUpdater(
        private val uploadSize: Long,
        private val totalSize: Long
    ) :
        Runnable {
        override fun run() {
            callback.onProgress((100 * uploadSize / totalSize).toInt())
        }

    }

    companion object {
        private const val DEFAULT_BUFFER_SIZE = 2048
    }
}
